### DISCUSSION TITLE
Running the Distributed Messenger example (OpenDDS-3.21)

### DISCUSSION THREAD

**Discussion by h3gui** [OpenDDS/OpenDDS#3683] [Q&A] [answered]
I do appreciate your help in running, under Windows 10 OS, the distributed messenger example as described in the getting_started document.
There was the assumption that a server is running (the following command -DCPSInfoRepo server07:12795). 
How to create it and use it?
If someone has run this example, I do appreciate the detailed sequence of steps.
Thank you

**Comment by mitza-oci** [comment]
Running the DCPSInfoRepo is described in the OpenDDS Developer's Guide section 2.1.7 (in "Getting Started").

**Comment by iguessthislldo** [comment, marked as answer]
If you're referring to `docs/getting_started.odt`, it introduces how to run the InroRepo earlier in the document, but those instructions are wrong now, specifically the part about `DCPSInfoRepo` being in `ACE_ROOT/bin`. This file hasn't been updated in a long time and probably should be removed.

Section 2.1.7 "Running The Example" in our the [OpenDDS Developer's Guide](https://download.objectcomputing.com/OpenDDS/OpenDDS-latest.pdf) gives information on the example in detail. The [Windows quick start](https://opendds.org/quickstart/GettingStartedWindows.html) linked in your earlier discussion also says how to run the example and if you ran `run_test.pl` according to those instructions it will print out the commands it's running.

**Reply by h3gui** [reply]
Thank you Adam and Fred. Running the Messenger on a single host (starting on section 2.1.7 of the OpenDDS Developer's Guide) works fine with the various methods described in the user guide as well as with run_test.pl. 
I wanted to go to the next step by having multiple connections (publishers and subscribers). The docs/getting_started.odt has a "Running Distributed Messenger" section with "Starting the InfoRepo" and "Starting the Publisher and Subscriber" as subsections. I am stuck at those.

**Reply by iguessthislldo** [reply]
There's probably a better way, but you could do that by running the commands you want in separate windows. The main issue you would run into is that the example isn't setup to expect multiple publishers or subscribers. They both wait for a single partner to be available, do their work and exit immediately after that.

Waiting for a particular setup of remote participants can be a little tricky depending on what you want to do, but I tried this just now and it seems to work for what I wanted to show you. Here I'm doing 2 subscribers and 2 publishers:

In `Publisher.cpp` change the 1 in this line to 2:

https://github.com/objectcomputing/OpenDDS/blob/5104ac052ac3811ddf49493c1d26dee8b63812dc/DevGuideExamples/DCPS/Messenger/Publisher.cpp#L130

In `Subscriber.cpp` change the `matches.total_count > 0` to `matches.total_count > 1`:

https://github.com/objectcomputing/OpenDDS/blob/5104ac052ac3811ddf49493c1d26dee8b63812dc/DevGuideExamples/DCPS/Messenger/Subscriber.cpp#L135

Then rebuild the example.

To run it, open five Command Prompts. They can be regular command prompts. In each one change to the OpenDDS folder, run `setenv`, and then `cd DevGuideExamples/DCPS/Messenger`. After that run one of these commands in each command prompt:

```
DCPSInfoRepo -o repo.ior
```

(The InfoRepo has to be run first. I don't think the order of the rest of the commands matter since they're waiting for all four to be present.)

```
publisher -DCPSDebugLevel 1
```
```
publisher -DCPSDebugLevel 1
```
```
subscriber -DCPSDebugLevel 1
```
```
subscriber -DCPSDebugLevel 1
```

Once you have both publishers and both subscribers running they should start. Each subscriber should get 20 messages each, two sets going from "count = 0" to "count = 9". They might interweave, for example mine went 0-5, 0-9, 6-9.

**Reply by h3gui** [reply]
Thank you very much Fred for the detailed steps that worked as written.
